[SDK-2825] InAppMessages using Flutter Streams (#3829)
braze-inc/braze-docs: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md

{% endtab %}
{% endtabs %}

## Receiving in-app message data

> To receive in-app message data in your Flutter app, `BrazePlugin` supports 2 methods for receiving in-app message data. The first method uses a data callback and the second method uses [Streams](https://dart.dev/tutorials/language/streams) (Preferred).

Author: Thanks Jared, I now updated the wording!

{% endtab %}
{% endtabs %}

## Receiving in-app message data

> To receive in-app message data in your Flutter app, the `BrazePlugin` supports receiving in-app message data using Dart [Streams](https://dart.dev/tutorials/language/streams) **(Recommended)** or by using a data callback **(Legacy)**.

Reviewer: ```suggestion
To receive in-app message data in your Flutter app, the `BrazePlugin` supports sending in-app message data using Dart [Streams](https://dart.dev/tutorials/language/streams) **(Recommended)** or by using a data callback **(Legacy)**.
```
Not sure if this is ideal, but imo the original reads a bit repetitive

Reviewer: (same comment on the iOS side)

> To receive in-app message data in your Flutter app, `BrazePlugin` supports two methods for receiving in-app message data. The first method uses Dart [Streams](https://dart.dev/tutorials/language/streams) **(Recommended)** and the second method uses a data callback **(Legacy)**.

Reviewer 2: ```suggestion
To receive in-app message data in your Flutter app, the `BrazePlugin` supports receiving in-app message data using Dart [Streams](https://dart.dev/tutorials/language/streams) **(Recommended)** or by using a data callback **(Legacy)**.
```

The `BrazeInAppMessage` object supports a subset of fields available in the native model objects, including `uri`, `message`, `header`, `buttons`, `extras`, and more.

> {% alert note %} The legacy data callback method is soon to be deprecated. Note that in-app messages can be added to both data streams and data callbacks. If you have already integrated data callbacks and wish to receive only data streams, remove any callback logic to ensure that in-app messages are processed exactly once. {% endalert %}

Reviewer 2: ```suggestion
{% alert note %} The legacy data callback method will soon be deprecated. Note that in-app messages can be added to both data streams and data callbacks. If you have already integrated data callbacks and wish to use data streams, remove any callback logic to ensure that in-app messages are processed exactly once. {% endalert %}
```

### Method 1: In-app message data streams (Recommended)

You can set a data stream listener in Dart to receive in-app message data in the Flutter host app.

To begin listening to the stream, use the code below to create a `StreamSubscription` in your Flutter app and call the `subscribeToInAppMessages()` method with a function that takes a `BrazeInAppMessage` instance. Remember to `cancel()` the stream subscription when it is no longer needed.

> ```dart

Reviewer: add space above

{% endtab %}
{% endtabs %}

#### Replaying the callback for in-app messages

To store any in-app messages triggered before the callback is available and replay them once it is set, add the following entry to the `customConfigs` map when intializing the `BrazePlugin`:
```dart
BrazePlugin braze = new BrazePlugin(customConfigs: {replayCallbacksConfigKey: true});
```

> ### Method 2: In-app message data streams (Preferred)

Reviewer 2: We should move this up as the main (and first) way to integrate receiving IAM data

I also think we can even remove the old callback method and only mention the streams integration to make sure new clients don't integrate with the (soon-to-be deprecated) callback method. I'm curious if others have the same thoughts for that

Reviewer 3: I think we should leave the callback method in the docs for now but streams should definitely go first and we can leave a not by the callback code that that integration is the legacy integration

Author: Thanks Daniel and Chava! I now updated the page to have the `Method 1: In-app message data streams (Recommended)` first and `Method 2: In-app message data callback (Legacy)` second. I also included an {% alert note %} to mention the legacy method is soon to be deprecated and when migrating to streams, they should remove any callback logic to ensure that an IAM is processed exactly once.

You can set a data stream listener in Dart to receive in-app message data in the Flutter host app.

> To begin listening to the stream, use the code below to create a `StreamSubscription` in your Flutter app and call the `listen()` method with a function that takes a `BrazeInAppMessage` instance. Remember to `cancel()` the stream subscription when no longer needed.

Reviewer 2: ```suggestion
To begin listening to the stream, use the code below to create a `StreamSubscription` in your Flutter app and call the `listen()` method with a function that takes a `BrazeInAppMessage` instance. Remember to `cancel()` the stream subscription when it is no longer needed.
```
